### PR TITLE
SimPPS/RPDigiProducer package for PPS Full simulation

### PR DESCRIPTION
**Pull request by mundim**
#### PR description:
New package with Digi producer for the strip sensor of PPS, needed for the full simulation

#### PR validation:

The following tests have been run:
scram b runtests
scram b code-checks
scram b code-formats (which triggered some style changes in the files)

runTheMatrix.py --job-reports -j 14 -s -l limited -i all --ibeos

which produced the output
32 30 28 23 12 2 1 1 1 tests passed, 0 1 1 0 0 0 0 0 0 failed

additional effort was done to track down issues that could trigger the same comments and in the previous PR related to the same package.

#### if this PR is a backport please specify the original PR:

Before submitting your pull requests, make sure you followed this checklist:
- verify that the PR is really intended for the chosen branch
- verify that changes follow [CMS Naming, Coding, And Style Rules](http://cms-sw.github.io/cms_coding_rules.html)
- verify that the PR passes the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html)

### CODE REVIEW

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
+1

**Comment by mundim** [comment]
Based on the comments from @fabiocos in the PR 27559, the code has been updated because it has a similar coding style (duplication of methods in a class instead of using the ones from LocalPoint class).

**Comment by civanch** [comment]
please test

**Comment by mundim** [comment]
Any news about this PR, please? thanks

File: SimPPS/RPDigiProducer/plugins/RPGaussianTailNoiseAdder.cc [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  verbosity_ = verbosity;
+  strips_above_threshold_prob_ = TMath::Erfc(theStripThresholdInE / sqrt(2.0) / theNoiseInElectrons) / 2;
+}
+
+simromanpot::strip_charge_map RPGaussianTailNoiseAdder::addNoise(const simromanpot::strip_charge_map &theSignal) {
+  simromanpot::strip_charge_map the_strip_charge_map;
+
+  // noise on strips with signal:
+  for (simromanpot::strip_charge_map::const_iterator i = theSignal.begin(); i != theSignal.end(); ++i) {
+    double noise = CLHEP::RandGauss::shoot(0.0, theNoiseInElectrons);
```
Comment (civanch) [inline SimPPS/RPDigiProducer/plugins/RPGaussianTailNoiseAdder.cc]: @mundim , normally CLHEP random engine should be defined in a producer and should be given by reference to all classes which sample digi. It is also possible if CLHEP random engine is initialized in the producer and use in this way (as it is done in hit production). Can you guarantee that the engine is properly initialized in the produce(...) method?

File: SimPPS/RPDigiProducer/plugins/RPGaussianTailNoiseAdder.cc [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    the_strip_charge_map[i->first] = i->second + noise;
+    if (verbosity_)
+      edm::LogInfo("RPDigiProducer") << "noise added to signal strips: " << noise << "\n";
+  }
+
+  // noise on the other strips
+  int strips_no_above_threshold = gRandom->Binomial(numStrips_, strips_above_threshold_prob_);
```
Comment (civanch) [inline SimPPS/RPDigiProducer/plugins/RPGaussianTailNoiseAdder.cc]: @mundim , how gRandom is initialized?

Comment (civanch) [reply SimPPS/RPDigiProducer/plugins/RPGaussianTailNoiseAdder.cc]: @mundim , there are equivalent CLHEP methods. Not necessary to use TRandom at all.

File: SimPPS/RPDigiProducer/plugins/RPGaussianTailNoiseAdder.cc (hunk 1 above)
Comment (mundim) [reply SimPPS/RPDigiProducer/plugins/RPGaussianTailNoiseAdder.cc]: Changes the constructor, passing the CLHEP::RandomEngine instanciated and inicialized by the producer.

File: SimPPS/RPDigiProducer/plugins/RPGaussianTailNoiseAdder.cc (hunk 2 above)
Comment (mundim) [reply SimPPS/RPDigiProducer/plugins/RPGaussianTailNoiseAdder.cc]: Changed to the CLHEP equivalent. It was also replaced the TMath::Erfc by the std::erfc as advised by Ianna at the time of the PPS Geometry PR.

**Comment by civanch** [comment]
please test

File: SimPPS/RPDigiProducer/plugins/BuildFile.xml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+<use   name="SimTracker/Common"/>
+<use   name="RecoCTPPS/PixelLocal"/>
```
Comment (civanch) [inline SimPPS/RPDigiProducer/plugins/BuildFile.xml]: @mundim , sorry I miss this before, dependence on RecoCTPPS is not needed.

Comment (mundim) [reply SimPPS/RPDigiProducer/plugins/BuildFile.xml]: I'm the one to be sorry. I have done a detailed analysis on the buildfile and this scaped my attention. it was needed to be replaced by other more "primitive" packages, like CondFormats, DataFormats, etc.

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
+1

File: SimPPS/RPDigiProducer/plugins/RPGaussianTailNoiseAdder.cc [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    if (verbosity_)
+      edm::LogInfo("RPDigiProducer") << "noise added to signal strips: " << noise << "\n";
+  }
+
+  // noise on the other strips
+  int strips_no_above_threshold =
+      CLHEP::RandBinomial::shoot(&(rndEngine_), (long)numStrips_, strips_above_threshold_prob_);
+
+  for (int j = 0; j < strips_no_above_threshold; j++) {
+    int strip = CLHEP::RandFlat::shootInt(numStrips_);
```
Comment (fabiocos) [inline SimPPS/RPDigiProducer/plugins/RPGaussianTailNoiseAdder.cc]: @mundim @civanch why this does not use the engine?

File: SimPPS/RPDigiProducer/plugins/RPVFATSimulator.cc [hunk 5]
```diff
@@ -0,0 +1,45 @@
+#include "SimPPS/RPDigiProducer/plugins/RPVFATSimulator.h"
+#include "Geometry/VeryForwardRPTopology/interface/RPTopology.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include <vector>
+#include "TRandom.h"
```
Comment (fabiocos) [inline SimPPS/RPDigiProducer/plugins/RPVFATSimulator.cc]: @mundim @civanch why this?

File: SimPPS/RPDigiProducer/test/RPDigiProducerTest_cfg.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+process.source = cms.Source("EmptySource")
+
+################## STEP 2 - process.generator
+
+# Use random number generator service
+process.load("Configuration.TotemCommon.RandomNumbers_cfi")
+
+# Monte Carlo gun - elastic specific
+energy = "7000"
+import IOMC.Elegent.ElegentSource_cfi
```
Comment (fabiocos) [inline SimPPS/RPDigiProducer/test/RPDigiProducerTest_cfg.py]: @mundim @civanch what is this? It does not look a CMSSW package

File: SimPPS/RPDigiProducer/plugins/RPVFATSimulator.cc (hunk 5 above)
Comment (mundim) [reply SimPPS/RPDigiProducer/plugins/RPVFATSimulator.cc]: Sorry, it was left behind when moving to CLHEP random generators. Solved.

File: SimPPS/RPDigiProducer/plugins/RPGaussianTailNoiseAdder.cc (hunk 4 above)
Comment (mundim) [reply SimPPS/RPDigiProducer/plugins/RPGaussianTailNoiseAdder.cc]: It scaped the attention when moving from TRandom generator to the CLHEP equivalent.

File: SimPPS/RPDigiProducer/test/RPDigiProducerTest_cfg.py (hunk 6 above)
Comment (mundim) [reply SimPPS/RPDigiProducer/test/RPDigiProducerTest_cfg.py]: It was a config test designed for totem environment. Changed accordingly, but it can not be run until all PPS simulation is merged because of dependencies on g4SimHit config file that is not yet in the release.

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
+1
@fabiocos , it seems, that all problems of this PR are now fixed

**Comment by fabiocos** [comment]
```
15:51 cmsdev25 2920> less cmsExec_RPDigiProducerTest_cfg_20190813_155114.log 
cmsExec: Starting job on  Tue Aug 13 15:51:14 CEST 2019

cmsExec: Running on  Linux cmsdev25.cern.ch 3.10.0-957.21.3.el7.x86_64 #1 SMP Tue Jun 18 16:35:19 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
cmsExec: System release  CentOS Linux release 7.6.1810 (Core)

cmsExec: CMSSW_RELEASE_BASE  /cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc700/cms/cmssw-patch/CMSSW_11_0_X_2019-08-12-2300
cmsExec: CMSSW_BASE          /build/fabiocos/110X/27608/CMSSW_11_0_X_2019-08-12-2300
cmsExec: working directory   /build/fabiocos/110X/27608/CMSSW_11_0_X_2019-08-12-2300/work

total 12
-rw-r--r--. 1 fabiocos zh 4144 Aug 13 15:50 RPDigiProducerTest_cfg.py
-rw-r--r--. 1 fabiocos zh  554 Aug 13 15:51 cmsExec_RPDigiProducerTest_cfg_20190813_155114.log

Begin processing the 1st record. Run 1, Event 1, LumiSection 1 on stream 0 at 13-Aug-2019 15:51:29.907 CEST
TVector3 A 3D physics vector (x,y,z)=(13.323109,1.691054,212460.000000) (rho,theta,phi)=(212460.000424,0.003622,7.233664)
TVector3 A 3D physics vector (x,y,z)=(0.037150,0.258219,6048.356623) (rho,theta,phi)=(6048.356629,0.002471,81.812979)
TVector3 A 3D physics vector (x,y,z)=(5.576807,1.020566,-212460.000000) (rho,theta,phi)=(212460.000076,179.998471,10.370482)
TVector3 A 3D physics vector (x,y,z)=(-0.882256,0.039799,-5840.064317) (rho,theta,phi)=(5840.064384,179.991336,177.417088)
----- Begin Fatal Exception 13-Aug-2019 15:51:40 CEST-----------------------
An exception of category 'PluginNotFound' occurred while
   [0] Processing  Event run: 1 lumi: 1 event: 1 stream: 0
   [1] Running path 'g4Simhits_step'
   [2] Calling method for module OscarMTProducer/'g4SimHits'
Exception Message:
Unable to find plugin 'CTPPSSensitiveDetector' in category 'CMS Simulation SensitiveDetectorFactory'. Please check spelling of name.
----- End Fatal Exception -------------------------------------------------

```

**Comment by fabiocos** [comment]
@mundim @civanch the problem does not seem directly in the digitizer code, but I wonder how this was tested...

**Comment by mundim** [comment]
> @mundim @civanch the problem does not seem directly in the digitizer code, but I wonder how this was tested...

As said in https://github.com/cms-sw/cmssw/pull/27608#discussion_r313036174, the test depends on a change in the g4SimHits config file, which is to be submitted in the next PR, including all changes in config files needed to really insert the PPS simulation into the cms simulation chain. All pieces have been tested in private simulations.

**Comment by civanch** [comment]
@fabiocos , it seems a bit early to this test. This PR is still a component.

**Comment by fabiocos** [comment]
@mundim @civanch ok, so let's move forward with the SIM part, and then come back to this. I do not understand how this was and could be tested

**Comment by fabiocos** [comment]
hold

**Comment by mundim** [comment]
> @mundim @civanch ok, so let's move forward with the SIM part, and then come back to this. I do not understand how this was and could be tested

sorry for the long delay in answering. I was thinking about how to do it, since this kind of comment is very hard to be replied. As mentioned at least twice, this piece of code can not be tested right away because it is missing other components.

Concerning the tests, it was indeed tested, as presented in the Simulation meeting of June, 8th, 2018 (see agenda https://indico.cern.ch/event/735560/).
Then we tried a PR with all components, which would allow testing without problems, but it was a very big PR (142 files). It was then canceled and splitted in several smaller PRs uppon request of the release management.

In summary, even after the other ongoing PR (27673) being merged, it will not be possible to test it because the last chunk of code (the configuration files) is still to be submitted (including the g4SimHit_cfi.py).

**Comment by kpedro88** [comment]
My conclusion is that #27673 and the followup PR with the config files should be merged before this one. It looks like the first one is ready.

**Comment by mundim** [comment]
> My conclusion is that #27673 and the followup PR with the config files should be merged before this one. It looks like the first one is ready.

Waiting for the 27673 to be merged before this one should not be a problem, but this one must go before the PR with the config files, otherwise, the tests will fail since they require this code.

**Comment by kpedro88** [comment]
It's possible to test two PRs together, so we can plan to do that once the config file PR is submitted and the other one is merged.

**Comment by mundim** [comment]
I was asked to submit one PR at a time, waiting the approval of the previous one to submit the next. So, I'll proceed with the next PR then. Thanks.

**Comment by fabiocos** [comment]
now #27673 has been merged, but the situation is unchanged. I believe that PPS is still not activated inside the master configuration of the simulation, @mundim @civanch do you confirm?

**Comment by mundim** [comment]
Yes, this is known. As told to Vladimir, the next PR should include all config files needed to activate PPS simulation. This was (apparently) the easiest way to have PPS included using smaller and independent PRs without disrupting the standard cmssw tests.

**Comment by fabiocos** [comment]
unhold

**Comment by mundim** [comment]
> @mundim @civanch I think that logically it makes sense to first have a full working chain at SIM level, produce hits, and then test and integrate the digi code. As it is now this PR cannot be used anyway.
> When is the PR activating the PPS in Geant4 expected?

A new PR was submitted (27990) to include PPS at sim level into cms simulation.

**Comment by fabiocos** [comment]
@mundim so far no event content is defined to use the digis produced, please correct me in case, when is this expected?

**Comment by fabiocos** [comment]
+1

**Comment by mundim** [comment]
@fabiocos As advised, we are trying to make orthogonal/small PR and one at a time, so now, that almost all pieces are merged, I'm going to issue the last PR with the config that will integrate all pieces together.
